File: 2017/day_12/ruby/day12.rb
class Graph
    attr_accessor :nodes

    def initialize
        @nodes = []
        @g = {}
    end

    def add_edge(start_vertex, end_vertex, weight = 1, undirected = true)
        # Add reciprocal edges to graph map with weight
        if @g.has_key?(start_vertex)
            @g[start_vertex][end_vertex] = weight
        else
            @g[start_vertex] = {end_vertex => weight}
        end
        # Undirected graph
        if @g.has_key?(end_vertex)
            @g[end_vertex][start_vertex] = weight
        else
            @g[end_vertex] = {start_vertex => weight}
        end

        # Add to nodes list
        @nodes << start_vertex unless @nodes.include?(start_vertex)
        # Undirected graph
        @nodes << end_vertex unless @nodes.include?(end_vertex)
    end

    def to_s
        str = []
        @nodes.each do |n|
            str << "node #{n} with neigbors " + @g[n].keys.join(", ")
        end
        str.join("\n")            
    end

    def read_from_file(source_file)
        # Read source file and strip line endings
        lines = []
        lines = IO.readlines(source_file)
        lines.map! {|line| line.strip}
        # For each line split the input
        lines.each do |line|
            lhs, rhs = line.split(" <-> ")
            neighbors = rhs.split(", ")
            neighbors.each do |node|
                add_edge(lhs, node)
            end
        end
    end

    def [](k)
        @g[k]
    end
end

class DFS
    attr_accessor :visited

    def initialize(graph, source_node)
        @graph = graph
        @source_node = source_node
        @visited = []
        @edge_to = {}

        dfs(source_node)
    end

    def dfs(node)
        @visited << node
        @graph[node].keys.each do |adj_node|
            next if @visited.include?(adj_node)
            dfs(adj_node)
            @edge_to[adj_node] = node
        end
    end

    def has_path_to?(target_node)
        @visited.include?(target_node)
    end

    def reachable_nodes
        @visited.uniq
    end
end

# Sample
g1 = Graph.new
g1.read_from_file("../sample.txt")
puts g1
dfs1 = DFS.new(g1, "0")

puts "Number of programs in group: #{dfs1.reachable_nodes.size}"
puts dfs1.reachable_nodes.join(", ")

# Part 1
g2 = Graph.new
g2.read_from_file("../input.txt")
puts g2
dfs2 = DFS.new(g2, "0")

puts "Number of programs in group: #{dfs2.reachable_nodes.size}"
puts dfs2.reachable_nodes.join(", ")

# Part 2
nodes = g2.nodes
components = 0
while nodes.size > 0
    source = nodes[0]
    dfs = DFS.new(g2, source)
    nodes = nodes - dfs.reachable_nodes
    components = components + 1
    puts "Component #{components}"
    puts dfs.reachable_nodes.join(", ")
end

puts "Total of #{components} components in Graph"

File: 2017/day_12/elixir/lib/day12.ex
defmodule Day12 do

  def read_input(source_file) do
    File.stream!(source_file)
    |> Enum.map(&String.trim/1)
    |> Enum.reduce(Graph.new(type: :undirected), &Day12.make_graph/2)
  end

  def make_graph(line, g = %Graph{}) do
    [source, neighbour_list] = String.split(line, ~r/ \<-\> /)
    neighbours = String.split(neighbour_list, ~r/, /)
    Enum.reduce(neighbours, g, fn (node, acc) -> Graph.add_edge(acc, source, node) end)
  end

end

# Sample
g1 = Day12.read_input("../sample.txt")
c01 = g1 |> Graph.components |> Enum.filter(fn x -> Enum.member?(x, "0") end) |> List.flatten
IO.puts "Component that includes \"0\" is"
IO.inspect c01
IO.puts "It has #{Enum.count(c01)} programs"
IO.puts "There are #{g1 |> Graph.components |> Enum.count} many components\n\n"

# Part 1 and Part 2
g2 = Day12.read_input("../input.txt")
c02 = g2 |> Graph.components |> Enum.filter(fn x -> Enum.member?(x, "0") end) |> List.flatten
IO.puts "Component that includes \"0\" is"
IO.inspect c02
IO.puts "It has #{Enum.count(c02)} programs"
IO.puts "There are #{g2 |> Graph.components |> Enum.count} many components\n\n"
